src/libs/feeds.js

```diff
@@ -0,0 +1,186 @@
+/**
+ * @file Manages Canvas and Portal feeds for caching purposes
+ * @module feeds
+ */
+
+const canvas = require(__dirname + '/canvas.js');
+const portal = require(__dirname + '/portal.js');
+const users  = require(__dirname + '/users.js');
+
+/**
+ * Update cached Canvas feed for a user
+ * @param {Object} db - Database object
+ * @param {string} user - Username
+ * @param {updateCanvasCacheCallback} callback - Callback
+ */
+
+/**
+ * Returns an error if any
+ * @callback updateCanvasCacheCallback
+ *
+ * @param {Object} err - Null if success, error object if failure
+ */
+
+function updateCanvasCache(db, user, callback) {
+	if(typeof callback !== 'function') return;
+
+	if(typeof db !== 'object') {
+		callback(new Error('Invalid database connection!'));
+		return;
+	}
+
+	const canvasdata = db.collection('canvasFeeds');
+
+	users.get(db, user, (err, isUser, userDoc) => {
+		if(err) {
+			callback(err);
+			return;
+		}
+		if(!isUser) {
+			callback(new Error('User doesn\'t exist!'));
+			return;
+		}
+
+		canvas.getFromURL(db, userDoc.user, (err, hasURL, events) => {
+			if(err) {
+				callback(err);
+				return;
+			}
+
+			canvasdata.deleteMany({ user: userDoc._id }, err => {
+				if(err) {
+					callback('There was an error removing the old events from the database!');
+					return;
+				}
+
+				events.forEach(e => e.user = userDoc._id);
+
+				canvasdata.insertMany(events, err => {
+					if(err) {
+						callback('There was an error inserting events into the database!');
+						return;
+					}
+
+					callback(null);
+				});
+			});
+		});
+	});
+}
+
+/**
+ * Add a user to the Portal queue
+ * @param {Object} db - Database object
+ * @param {string} user - Username
+ * @param {addPortalQueueCallback} callback - Callback
+ */
+
+/**
+ * Returns an error if any
+ * @callback addPortalQueueCallback
+ *
+ * @param {Object} err - Null if success, error object if failure
+ */
+
+function addPortalQueue(db, user, callback) {
```

Comment:
To help clarify, what I'm basically saying is that the scope of `addPortalQueue()` should be a little bit bigger. Here's how I think it would be best:
```
1. Get portal calendar
2. Does portal error?
    Yes: Add to queue (or remain in queue if already in it)
    No: Cache portal feed! We're all done! Also ensure user is not in the queue.
```
Then, you can simplify the `processPortalQueue()` a bit by simply calling `addPortalQueue()` for every person currently in the queue.